lib/chef/resource.rb

```diff
@@ -871,23 +864,197 @@ def after_created
       nil
     end
 
-    #
-    # The module where Chef should look for providers for this resource.
-    # The provider for `MyResource` will be looked up using
-    # `provider_base::MyResource`.  Defaults to `Chef::Provider`.
-    #
-    # @param arg [Module] The module containing providers for this resource
-    # @return [Module] The module containing providers for this resource
-    #
-    # @example
-    #   class MyResource < Chef::Resource
-    #     provider_base Chef::Provider::Deploy
-    #     # ...other stuff
-    #   end
-    #
-    def self.provider_base(arg=nil)
-      @provider_base ||= arg
-      @provider_base ||= Chef::Provider
+    # Provider lookup and naming
+    class<<self
+      #
+      # The DSL name of this resource (e.g. `package` or `yum_package`)
+      #
+      # @return [String] The DSL name of this resource.
+      #
+      # @deprecated Use resource_name instead.
+      #
+      def dsl_name
+        Chef::Log.deprecation "Resource.dsl_name is deprecated and will be removed in Chef 11.  Use resource_name instead."
+        if name
+          name = self.name.split('::')[-1]
+          convert_to_snake_case(name)
+        end
+      end
+
+      #
+      # The display name of this resource type, for printing purposes.
+      #
+      # @return [Symbol] The name of this resource type (e.g. `:execute`).
+      #
+      attr_accessor :resource_name
+      def resource_name(value=NULL_ARG)
+        if value != NULL_ARG
+          self.resource_name = value.to_sym
```

Comment:
This looks like `nil` can be passed in as the resource name, which produces the dread `NoMethodError: undefined method `to_sym' for nil:NilClass` error that we should want to avoid.